compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -2861,7 +2861,39 @@ class Typer extends Namer
   final def adapt(tree: Tree, pt: Type)(using Context): Tree =
     adapt(tree, pt, ctx.typerState.ownedVars)
 
-  private def adapt1(tree: Tree, pt: Type, locked: TypeVars, tryGadtHealing: Boolean)(using Context): Tree = {
+  // See https://github.com/lampepfl/dotty/issues/9203 for why this is needed
+  // So far it is guarding only `adapt1` since that's the highest known entry point
+  // stack call where the detection for the debugging of #9203 is possible.
+  private inline def spaceComplexityTracker[T](tree: Tree)(task: => T)(using Context) =
+    def size = ctx.base.uniquesSizes.values
+      .map { case (size, _, _) => size }.sum
+    val before = size
+    val res = task
+    val after = size
+    val delta = after - before
+    if delta > 100000
+      println(  // ctx.warning does not output the message in this place, probably waits
```

Comment:
Why not emit an error as we do in [Inliner.scala#L132](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/typer/Inliner.scala#L132)? There we return an `errorTree` to avoid further expansion, this may also work here.